src/Titon/Console/StyleDefinition.hh

```diff
@@ -0,0 +1,227 @@
+        'light_red'    => '1;31',
+        'purple'       => '0;35',
+        'light_purple' => '1;35',
+        'brown'        => '0;33',
+        'yellow'       => '1;33',
+        'light_gray'   => '0;37',
+        'white'        => '1;37',
+    };
+
+    /**
+     * The various effects to apply.
+     *
+     * @var Vector<string>
+     */
+    private Vector<string> $effectsList;
+
+    /**
+     * Data structure containing available effects.
+     *
+     * @var Map<string, int>
+     */
+    protected Map<string, int> $effectsMap = Map {
+        'defaults'  => 0,
+        'bold'      => 1,
+        'underline' => 4,
+        'blink'     => 5,
+        'reverse'   => 7,
+        'conceal'   => 8,
+    };
+
+    /**
+     * Create a new `StyleDefinition`.
+     *
+     * @param string $fgColor           The foreground color of the style
+     * @param string $bgColor           The background color of the style
+     * @param Vector<string> $effects   The effects of the style
+     */
+    public function __construct(string $fgColor, string $bgColor = '', Vector<string> $effects = Vector {}) {
+        $this->fgColor = $fgColor;
+        $this->bgColor = $bgColor;
+        $this->effectsList = $effects;
+    }
+
+    /**
+     * Format the contents between the given XML tag with the style definition.
+     *
+     * @param string $xmlTag    The XML tag
+     * @param string $value     The contents to format
+     *
+     * @return string
+     */
+    public function format(string $xmlTag, string $value): string {
+        $values = $this->getValueBetweenTags($xmlTag, $value);
+        $retval = $value;
+        foreach ($values as $val) {
+            $valueReplaced = '<' . $xmlTag . '>' . $val . '</' . $xmlTag . '>';
+            $valueResult = $this->replaceTagColors($val);
+
+            $retval = str_replace($valueReplaced, $valueResult, $retval);
+        }
+
+        return $retval;
+    }
+
+    /**
+     * Retrieve the background color code of the `StyleDefinition`.
+     *
+     * @return string
+     */
+    public function getBgColorCode(): string {
+        if (!is_null($bgColor = $this->bgColorsMap->get($this->bgColor))) {
+            invariant(is_string($bgColor), "Must be a string.");
+
+            return $bgColor;
+        }
+
+        return '';
+    }
+
+    /**
+     * Retrieve the effects code of the `StyleDefinition`.
+     *
+     * @return string
+     */
+    public function getEffectCode(string $effect): int {
+        return $this->effectsMap[$effect];
+    }
+
+    /**
+     * Retrieve the code to end the `StyleDefinition`.
+     *
+     * @return string
+     */
+    public function getEndCode(): string {
+        return "\033[0m";
+    }
+
+    /**
+     * Retrieve the foreground color code of the `StyleDefinition`.
+     *
+     * @return string
+     */
+    public function getFgColorCode(): string {
+        if (!is_null($fgColor = $this->fgColorsMap->get($this->fgColor))) {
```

Comment:
Are these checks necessary? Could probably just do `return (string) $this->fgColorsMap->get($this->fgColor);`.